### PR TITLE
Rewrite `test_determinism`; closes #6882

### LINKED ISSUE #6882: SVG backend: test_determinism
I do not like the fact that it stores temporary files in the cwd (it is ugly way and unrandomized file names are not thread safe).

Possible solutions:
- Use `TemporaryFile`
- Use stdout, pipe or shared memory for interprocess communication
- Just make all the checks in the tests process

If I understand correctly the purpose of the subprocess call -- it is used to check if the outputs of multiple python runs are the same, but they will likely be the same, because the single source of difference is a random seed which is disabled by `-R` key to the python interpreter. I think that three consecutive calls to `savefig` will do the same job as `_test_determinism` (with the help of `_test_determinism_save`) does currently.

### PR DESCRIPTION
**Pull request by Kojoley**
Changes:
  - uses pipe instead of temporary file
  - checks if test really did something useful
    (dict keys must have different order in subprocesses)

The test can fail if somehow the child processes have the same `PYTHONHASHSEED` value. This should not happen often (or ever happen), but there is a little chance that Python interpreter can generate the same value in both processes. Solution for this can be following: set `PYTHONHASHSEED` value for processes manually before running them, but I don't know if it is a good choice or not, so I made PR without it.

### CODE REVIEW

File: lib/matplotlib/tests/test_backend_svg.py [hunk 1]
```diff
@@ -150,48 +152,78 @@ def _test_determinism_save(filename, usetex):
...
 
 
-def _test_determinism(filename, usetex):
-    import os
-    import sys
-    from subprocess import check_output, STDOUT, CalledProcessError
-    from nose.tools import assert_equal
-    plots = []
+def _test_determinism_helper(pipe, usetex):
+    guard = list({str(i): i for i in range(10)})
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_backend_svg.py]: Running this line, I get what it does, but it's going to need a comment lest someone "optimize" it into a more straightforward list comprehension.

Comment (Kojoley) [inline lib/matplotlib/tests/test_backend_svg.py]: Agreed

**Comment by Kojoley** [comment]
I cannot understand why it fails on travis on all builds except 3.4 (It works perfectly with 2.7 and 3.5 locally). I am digging this down, but still have no luck.

**Comment by WeatherGod** [comment]
I am still trying to wrap my head around this, but isn't it fundamentally foolhardy to test that the dictionary keys are in different order? While it would *likely* be different, it isn't *guaranteed* to be different, or am I just misunderstanding all of this?

Also, I am not sure the setting/resetting of PYTHONHASHSEED is correct, but I haven't looked closely enough. In particular, it doesn't seem to be safe from race conditions.

**Comment by Kojoley** [comment]
`os.environ` is thread-safe since 2.7, but I have added `PYTHONHASHSEED` magic because of travis failures, in trying to understand where is the problem.

**Comment by WeatherGod** [comment]
os.environ thread-safe? really? Not in my experience. I personally fixed a
bug in a co-worker's wsgi app where he effectively was storing a path into
an environment variable and retrieving it later in the code, and ran into
all sorts of problems when two different functions were doing that to the
same environment variable. All I did was eliminate the usage of environment
variables and all race-conditions in his wsgi app stopped.

**Comment by Kojoley** [comment]
`os.environ` is not shared memory here. I do not access here it from multiple threads, because I even do not have them here. Every child process have their own `os.environ` which is copied on the process start.

After some investigation I have remembered that multiprocessing module on POSIX system uses forking, so here is the source of the problem with the same dictionary order in child processes on Linux. There is no solution or workaround for it, as internal hash salt (`_Py_HashSecret`) initialized only once at the startup and cannot be changed (due to limitations in the interpreter implementation).

**Pull request closed without merging**